axum-macros/src/debug_handler.rs

```diff
@@ -1,31 +1,64 @@
 use proc_macro2::TokenStream;
 use quote::{format_ident, quote, quote_spanned};
-use syn::{parse::Parse, spanned::Spanned, FnArg, ItemFn, Token, Type};
+use syn::{
+    parse::Parse, punctuated::Punctuated, spanned::Spanned, FnArg, GenericParam, ItemFn, Token,
+    Type,
+};
+
+use self::specializer::Specializer;
+
+mod specializer;
 
 pub(crate) fn expand(attr: Attrs, item_fn: ItemFn) -> TokenStream {
+    // TODO error on const generics and liftetimes
+    // TODO error on not all generic params specified using with
+
+    let generic_param_idents = item_fn
+        .sig
+        .generics
+        .params
+        .iter()
+        .enumerate()
+        .flat_map(|(_idx, param)| {
+            match param {
+                GenericParam::Type(t) => Some(t.ident.clone()),
+                _ => {
+                    // TODO should we flag an error here
+                    None
+                }
+            }
+        })
+        .collect::<Vec<_>>();
+
+    let specializer = Specializer::new(generic_param_idents, attr.compute_specializations());
+
     let check_extractor_count = check_extractor_count(&item_fn);
     let check_request_last_extractor = check_request_last_extractor(&item_fn);
     let check_path_extractor = check_path_extractor(&item_fn);
     let check_multiple_body_extractors = check_multiple_body_extractors(&item_fn);
-    let check_output_impls_into_response = check_output_impls_into_response(&item_fn);
-
-    // If the function is generic, we can't reliably check its inputs or whether the future it
-    // returns is `Send`. Skip those checks to avoid unhelpful additional compiler errors.
-    let check_inputs_and_future_send = if item_fn.sig.generics.params.is_empty() {
+    let check_output_impls_into_response = check_output_impls_into_response(&item_fn, &specializer);
+
+    // If the function is generic and no with statement was provided, we can't reliably check its
+    // inputs or whether the future it returns is `Send`. Skip those checks to avoid unhelpful additional
+    // compiler errors.
+    let check_inputs_and_future_send = if item_fn.sig.generics.params.is_empty()
+        || attr.with_tys.is_some()
+    {
         let check_inputs_impls_from_request =
-            check_inputs_impls_from_request(&item_fn, &attr.body_ty);
-        let check_future_send = check_future_send(&item_fn);
-
+            check_inputs_impls_from_request(&item_fn, &attr.body_ty, &specializer);
+        let check_future_send = check_future_send(&item_fn, &specializer);
         quote! {
             #check_inputs_impls_from_request
             #check_future_send
         }
     } else {
         syn::Error::new_spanned(
-            &item_fn.sig.generics,
-            "`#[axum_macros::debug_handler]` doesn't support generic functions",
-        )
-        .into_compile_error()
+                &item_fn.sig.generics,
+                "`#[axum_macros::debug_handler]` use with 'with' attribute to support debugging generic functions",
```

Comment:
open for suggestions on better error message

nit: I will get rid of the double-with